qiskit/transpiler/passmanager.py

```diff
@@ -312,3 +312,179 @@ def passes(self) -> List[Dict[str, BasePass]]:
+    built using predefined stages:
+
+    1. Init - any initial passes that are run before we start embedding the circuit to the backend
+    2. Layout - This stage runs layout and maps the virtual qubits in the
+        circuit to the physical qubits on a backend
+    3. Routing - This stage runs after a layout has been run and will insert any
+        necessary gates to move the qubit states around until it can be run on
+        backend's compuling map.
+    4. Translation - Perform the basis gate translation, in other words translate the gates
+        in the circuit to the target backend's basis set
+    5. Pre-Optimization - Any passes to run before the main optimization loop
+    6. Optimization - The main optimization loop, this will typically run in a loop trying to optimize
+        the circuit until a condtion (such as fixed depth) is reached.
+    7. Post-Optimization - Any passes to run after the main optimization loop
+    8. Scheduling - Any hardware aware scheduling passes
+
+    These stages will be executed in order and any stage set to ``None`` will be skipped. If
+    a :class:`~qiskit.transpiler.PassManager` input is being used for more than 1 stage here
+    (for example in the case of a Pass that covers both Layout and Routing) you will want to set
+    that to the earliest stage in sequence that it covers.
+    """
+
+    phases = [
+        "init",
+        "layout",
+        "routing",
+        "translation",
+        "pre_optimization",
+        "optimization",
+        "post_optimization",
+        "scheduling",
+    ]
+
+    def __init__(
+        self,
+        init=None,
+        layout=None,
+        routing=None,
+        translation=None,
+        pre_optimization=None,
+        optimization=None,
+        post_optimization=None,
+        scheduling=None,
+    ):
+        """Initialize a new StructuredPassManager object
+
+        Args:
+            init (PassManager): A passmanager to run for the initial stage of the
+                compilation.
+            layout (PassManager): A passmanager to run for the layout stage of the
+                compilation.
+            routing (PassManager): A pass manager to run for the routing stage
+                of the compilation
+            translation (PassManager): A pass manager to run for the translation
+                stage of the compilation
+            pre_opt (PassManager): A pass manager to run before the optimization
+                loop
+            optimization (PassManager): A pass manager to run for the
+                optimization loop stage
+            post_opt (PassManager): A pass manager to run after the optimization
+                loop
+            scheduling (PassManager): A pass manager to run any scheduling passes
+        """
+        super().__init__()
+        self._init = init
+        self._layout = layout
+        self._routing = routing
+        self._translation = translation
+        self._pre_optimization = pre_optimization
```

Comment:
Do there really need to be so many stages hard coded? Skimming through the comments here everyone has their own ideas for what stages they want, so I think it would be better to make this completely configurable.

You could have something like
```
def __init__(self, pass_stages: List[Tuple[str, PassManager]])
```
where you can initialize with whatever stages you want, with this just being one example done like:
```
[("init", init), ("layout", layout), ...]
```
Then instead of the current properties for each stage you could have a generic method like:
```
def stage_pass_manager(self, stage: Union[str, int]) -> PassManager
```
to return a given pass manager stage based on label or position that you could append/modify (and another method to replace the pass manager for a stage).

You could also add an API for adding or removing stages after initialization, and a method to return the tuple of stage names like `def stages(self) -> Tuple[str]`

Reply:
I went this direction in: https://github.com/Qiskit/qiskit-terra/pull/6403/commits/aa1db1f8941cad54ed91ded539465a259640edaf I didn't use the tuple input but made stages a list of strings and then let users optionally define the initial value with a kwarg with that stage name.